### PR TITLE
Allow forward slash prefix to container name when running docker rm

### PR DESCRIPTION
**Pull request by brahmaroutu**
Addresses #6898
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

**Comment by duglin** [comment]
@brahmaroutu if its safe to remove the leading / then would it make sense to do this on the daemon side rather than the cli?  Then no matter what CLI is talking to the deamon (or even if someone is just using the REST APIs), we'll always catch this and remove it for them in one spot - instead of each CLI having to do it themselves.

**Comment by brahmaroutu** [comment]
@duglin This is an issue on the cli side, 
docker rm /abc will turn into a http request 
http://host:port/containers//abc ... and the "//" is causing the issue.

**Comment by duglin** [comment]
@brahmaroutu ok - as long as there isn't some additional logic/check that we can/should do on the daemon to help.

**Comment by brahmaroutu** [comment]
@duglin no need for other checks on daemon side. 
as mentioned above "docker rm /abc" does not call code on the daemon because the url is not formed correctly and the call does not go to the daemon.

File: integration-cli/docker_cli_rm_test.go [hunk 1]
```diff
@@ -140,6 +140,22 @@ func TestRmInvalidContainer(t *testing.T) {
...
 
+func TestRmContaineriNameWithSlash(t *testing.T) {
+	defer deleteAllContainers()
+
+	cmd := exec.Command(dockerBinary, "run", "--name", "testContainer", "busybox", "true")
+	if _, err := runCommand(cmd); err != nil {
+		t.Fatal(err)
+	}
+
+	cmd = exec.Command(dockerBinary, "rm", "-v", "/testContainer")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_rm_test.go]: Why the `-v`?

**Comment by cpuguy83** [comment]
It would be great to have this handled for other actions as well

File: integration-cli/docker_cli_rm_test.go (hunk 1 above)
Comment (brahmaroutu) [inline integration-cli/docker_cli_rm_test.go]: Oops cut&paste error. Thanks.

**Comment by brahmaroutu** [comment]
I will add this fix to all the commands that require this ex: pause, unpause, top, kill, etc But I want to do smaller PRs, it is easy to verify.

**Comment by icecrime** [comment]
Thanks @brahmaroutu! Unfortunately this needs a (painful) rebase as all clients commands have been split in many files :-(

**Comment by brahmaroutu** [comment]
@icecrime rebased the code accordingly.

File: api/client/rm.go [hunk 2]
```diff
@@ -30,6 +30,7 @@ func (cli *DockerCli) CmdRm(args ...string) error {
 
 	var encounteredError error
 	for _, name := range cmd.Args() {
+		name = stripPrefix(name)
```
Comment (icecrime) [inline api/client/rm.go]: Should we use [`strings.TrimPrefix`](http://golang.org/pkg/strings/#TrimPrefix) here?

Comment (jfrazelle) [inline api/client/rm.go]: good catch

Comment (brahmaroutu) [inline api/client/rm.go]: I added a function stripPrefix in utils.go

Comment (jfrazelle) [inline api/client/rm.go]: but go has builtin

Comment (brahmaroutu) [inline api/client/rm.go]: Yep, I agree and I can change. Only reason for the function is that all api calls using container name are going to use it. Just paranoia to see if another preprocessing may happen in the future. Let me know if you want to to change it in place.

Comment (icecrime) [inline api/client/rm.go]: Sorry, I don't get the argument for having our own function, and *even* if we have a good reason for that I believe it should be implemented in terms of `strings.TrimPrefix`.

**Comment by brahmaroutu** [comment]
@icecrime Please review.

**Comment by icecrime** [comment]
Testing this thing around, I'm quite unsure if that's the proper fix or if we should rather forbid leading `/` in container names, urlencode the name on the client-side before sending, or anything else...

Anyway, this patch is small enough and closes an issue so LGTM.

**Comment by unclejack** [comment]
The right fix for that issue is to disallow `/` in the start of the name of the container.
I'm sorry, I've just seen this PR now.

NOT LGTM

**Comment by cpuguy83** [comment]
@unclejack Currently `docker inspect` shows the "/" in the Name field, which is probably where this is a problem.
Maybe just remove it from that output?

**Comment by icecrime** [comment]
Alright back to design review, sorry for the false hope.

**Comment by stevvooe** [comment]
-1 on this. The container name should be validated before issuing any particular command. Names of the format "/foo" are not valid.

**Comment by icecrime** [comment]
Thanks for the reviews.

@brahmaroutu I'm closing this, but please feel free to submit a new PR with the appropriate fix, or ask us here to reopen before force-pushing the branch!

**Pull request closed without merging**